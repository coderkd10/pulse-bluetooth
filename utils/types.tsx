import { Peripheral, PeripheralInfo } from "react-native-ble-manager";

export type BTNotConnectedDevice = {
    peripheral: Peripheral,
    isConnecting: boolean,
};

export type BTConnectedDevice = {
    peripheral: Peripheral,
    peripheralInfo?: PeripheralInfo,
};

export type BTReadingState = {
    status: "reading", 
} | {
    status: "success",
    data: string,
} | {
    status: "error",
    error: string,
};

export type BTReadingCharacteristic = {
    serviceID: string,
    characteristicID: string,
    readingState: BTReadingState,
};

export type BTReadingDevice = {
    device: { id: string, name?: string },
    readableCharacteristics: Map<string, BTReadingCharacteristic>,
};
