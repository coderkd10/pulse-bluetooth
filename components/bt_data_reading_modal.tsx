import { Modal, Text, View, FlatList, Button, StyleSheet } from "react-native";
import { BTReadingDevice, BTReadingCharacteristic } from "@/utils/types";
import { DeviceName } from "./device_name";

type BTDataReadingModalProps = {
  currentReadingDevice: BTReadingDevice | null,
  handleClose: () => void,

};
export const BTDataReadingModal = ({ currentReadingDevice, handleClose }: BTDataReadingModalProps) => {
  if (!currentReadingDevice) {
    return <></>;
  }
  return (
    <Modal
      transparent={true}
      animationType="slide"
      visible={!!currentReadingDevice}
      onRequestClose={handleClose}
    >
      <View style={styles.modalContainer}>
        <View style={styles.modalContent}>
          <Text style={styles.modalTitle}>Reading Bluetooth Data</Text>
          <View>
            <Text style={styles.deviceNameTitle}>Device</Text>
            <View style={styles.deviceNameContainer}>
              <DeviceName id={currentReadingDevice.device.id} name={currentReadingDevice.device.name}/>
            </View>
          </View>
          <Text style={{ marginBottom: 10 }}>Found {currentReadingDevice.readableCharacteristics.size} readable data elements ("characteristics").</Text>
          <FlatList
            data={[...currentReadingDevice.readableCharacteristics.values()]}
            renderItem={renderListItem}
            keyExtractor={item => `${item.serviceID}-${item.characteristicID}`}
            style={styles.list}
          />
          <Button title="Done" onPress={handleClose} />
        </View>
      </View>
    </Modal>
  );
}

const renderListItem = ({item}: { item: BTReadingCharacteristic }) => {
  return (
    <View style={styles.listItemContainer}>
      <Text>
        <Text style={{ fontWeight: 'bold' }}>ServiceID: </Text>
        {item.serviceID}
      </Text>
      <Text style={{ marginBottom: 5 }}>
        <Text style={{ fontWeight: 'bold' }}>CharacteristicID: </Text>
        {item.characteristicID}
      </Text>
      {item.readingState.status === "reading" ? 
        <Text>Reading Data ...</Text>:
        item.readingState.status === "success" ?
          <Text>Read Data: {item.readingState.data}</Text> :
          <Text>Error: {item.readingState.error}</Text>
      }
    </View>
  );
}

const styles = StyleSheet.create({
  modalContainer: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: 'rgba(0, 0, 0, 0.7)'
  },
  modalContent: {
    width: '90%',
    height: '90%',
    padding: 20,
    backgroundColor: 'white',
    borderRadius: 10,
  },
  modalTitle: {
    fontSize: 24,
    fontWeight: 'bold',
    textAlign: 'center',
    marginBottom: 20,
  },
  modalCancelButton: {
    marginTop: 10,
  },
  modalCancelButtonText: {
    color: 'red',
    fontSize: 16,
  },
  deviceNameTitle: { 
    borderWidth: 1, borderBottomWidth: 0, borderTopLeftRadius: 4, borderTopRightRadius: 4,
    backgroundColor: '#999',
    color: '#fff',
    textAlign: 'center',
    padding: 2,
    fontWeight: 'bold',
  },
  deviceNameContainer: { 
    borderWidth: 1,
    padding: 12,
    marginBottom: 20,
  },
  list: {
    padding: 5,
    paddingBottom: 100,
    marginBottom: 20,
    borderWidth: 1,
    borderRadius: 5,
    minHeight: '10%',
  },
  listItemContainer: {
    padding: 10,
    marginBottom: 10,
    borderWidth: 1,
    borderColor: '#ccc',
    borderRadius: 5,
  }
});
