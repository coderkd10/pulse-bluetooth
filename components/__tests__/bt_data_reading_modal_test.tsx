import * as React from 'react';
import { render } from '@testing-library/react-native';

import { BTDataReadingModal } from '../bt_data_reading_modal';
import { BTReadingDevice } from '@/utils/types';

export const mockCurrentlyReadingDevice: BTReadingDevice = {
    device: {
        id: "62:64:0D:5F:87:9C",
        name: "JBL Flip",
    },
    readableCharacteristics: new Map([
        ["1800:2a00", {
            serviceID: "1800",
            characteristicID: "2a00",
            readingState: {
                status: "reading",
            },
        }],
        ["1800:2a01", {
            serviceID: "1800",
            characteristicID: "2a01",
            readingState: {
                status: "success",
                data: "aVBob25l",
            },
        }],
        ["89d3502b-0f36-433a-8ef4-c502ad55f8dc:c6b2f38c-23ab-46d8-a6ab-a3a870bbd5d7", {
            serviceID: "89d3502b-0f36-433a-8ef4-c502ad55f8dc",
            characteristicID: "c6b2f38c-23ab-46d8-a6ab-a3a870bbd5d7",
            readingState: {
                status: "error",
                error: "Error reading c6b2f38c-23ab-46d8-a6ab-a3a870bbd5d7 status=137",
            },
        }]
    ]),
};


it(`renders correctly`, () => {
    const tree = render(
        <BTDataReadingModal
            currentReadingDevice={mockCurrentlyReadingDevice}
            handleClose={jest.fn()}
        />
    ).toJSON();
  
    expect(tree).toMatchSnapshot();
});
