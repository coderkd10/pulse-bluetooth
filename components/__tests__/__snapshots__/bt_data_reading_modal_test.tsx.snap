// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Modal
  animationType="slide"
  hardwareAccelerated={false}
  onRequestClose={[MockFunction]}
  transparent={true}
  visible={true}
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "rgba(0, 0, 0, 0.7)",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <View
      style={
        {
          "backgroundColor": "white",
          "borderRadius": 10,
          "height": "90%",
          "padding": 20,
          "width": "90%",
        }
      }
    >
      <Text
        style={
          {
            "fontSize": 24,
            "fontWeight": "bold",
            "marginBottom": 20,
            "textAlign": "center",
          }
        }
      >
        Reading Bluetooth Data
      </Text>
      <View>
        <Text
          style={
            {
              "backgroundColor": "#999",
              "borderBottomWidth": 0,
              "borderTopLeftRadius": 4,
              "borderTopRightRadius": 4,
              "borderWidth": 1,
              "color": "#fff",
              "fontWeight": "bold",
              "padding": 2,
              "textAlign": "center",
            }
          }
        >
          Device
        </Text>
        <View
          style={
            {
              "borderWidth": 1,
              "marginBottom": 20,
              "padding": 12,
            }
          }
        >
          <View>
            <Text
              style={
                {
                  "fontSize": 16,
                }
              }
            >
              JBL Flip
            </Text>
            <Text
              style={
                {
                  "color": "#666",
                  "fontWeight": "bold",
                }
              }
            >
              62:64:0D:5F:87:9C
            </Text>
          </View>
        </View>
      </View>
      <Text
        style={
          {
            "marginBottom": 10,
          }
        }
      >
        Found 
        3
         readable data elements ("characteristics").
      </Text>
      <RCTScrollView
        data={
          [
            {
              "characteristicID": "2a00",
              "readingState": {
                "status": "reading",
              },
              "serviceID": "1800",
            },
            {
              "characteristicID": "2a01",
              "readingState": {
                "data": "aVBob25l",
                "status": "success",
              },
              "serviceID": "1800",
            },
            {
              "characteristicID": "c6b2f38c-23ab-46d8-a6ab-a3a870bbd5d7",
              "readingState": {
                "error": "Error reading c6b2f38c-23ab-46d8-a6ab-a3a870bbd5d7 status=137",
                "status": "error",
              },
              "serviceID": "89d3502b-0f36-433a-8ef4-c502ad55f8dc",
            },
          ]
        }
        getItem={[Function]}
        getItemCount={[Function]}
        keyExtractor={[Function]}
        onContentSizeChange={[Function]}
        onLayout={[Function]}
        onMomentumScrollBegin={[Function]}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        removeClippedSubviews={false}
        renderItem={[Function]}
        scrollEventThrottle={0.0001}
        stickyHeaderIndices={[]}
        style={
          {
            "borderRadius": 5,
            "borderWidth": 1,
            "marginBottom": 20,
            "minHeight": "10%",
            "padding": 5,
            "paddingBottom": 100,
          }
        }
        viewabilityConfigCallbackPairs={[]}
      >
        <View>
          <View
            onFocusCapture={[Function]}
            onLayout={[Function]}
            style={null}
          >
            <View
              style={
                {
                  "borderColor": "#ccc",
                  "borderRadius": 5,
                  "borderWidth": 1,
                  "marginBottom": 10,
                  "padding": 10,
                }
              }
            >
              <Text>
                <Text
                  style={
                    {
                      "fontWeight": "bold",
                    }
                  }
                >
                  ServiceID: 
                </Text>
                1800
              </Text>
              <Text
                style={
                  {
                    "marginBottom": 5,
                  }
                }
              >
                <Text
                  style={
                    {
                      "fontWeight": "bold",
                    }
                  }
                >
                  CharacteristicID: 
                </Text>
                2a00
              </Text>
              <Text>
                Reading Data ...
              </Text>
            </View>
          </View>
          <View
            onFocusCapture={[Function]}
            onLayout={[Function]}
            style={null}
          >
            <View
              style={
                {
                  "borderColor": "#ccc",
                  "borderRadius": 5,
                  "borderWidth": 1,
                  "marginBottom": 10,
                  "padding": 10,
                }
              }
            >
              <Text>
                <Text
                  style={
                    {
                      "fontWeight": "bold",
                    }
                  }
                >
                  ServiceID: 
                </Text>
                1800
              </Text>
              <Text
                style={
                  {
                    "marginBottom": 5,
                  }
                }
              >
                <Text
                  style={
                    {
                      "fontWeight": "bold",
                    }
                  }
                >
                  CharacteristicID: 
                </Text>
                2a01
              </Text>
              <Text>
                Read Data: 
                aVBob25l
              </Text>
            </View>
          </View>
          <View
            onFocusCapture={[Function]}
            onLayout={[Function]}
            style={null}
          >
            <View
              style={
                {
                  "borderColor": "#ccc",
                  "borderRadius": 5,
                  "borderWidth": 1,
                  "marginBottom": 10,
                  "padding": 10,
                }
              }
            >
              <Text>
                <Text
                  style={
                    {
                      "fontWeight": "bold",
                    }
                  }
                >
                  ServiceID: 
                </Text>
                89d3502b-0f36-433a-8ef4-c502ad55f8dc
              </Text>
              <Text
                style={
                  {
                    "marginBottom": 5,
                  }
                }
              >
                <Text
                  style={
                    {
                      "fontWeight": "bold",
                    }
                  }
                >
                  CharacteristicID: 
                </Text>
                c6b2f38c-23ab-46d8-a6ab-a3a870bbd5d7
              </Text>
              <Text>
                Error: 
                Error reading c6b2f38c-23ab-46d8-a6ab-a3a870bbd5d7 status=137
              </Text>
            </View>
          </View>
        </View>
      </RCTScrollView>
      <View
        accessibilityRole="button"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            [
              {},
            ]
          }
        >
          <Text
            style={
              [
                {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "margin": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            Done
          </Text>
        </View>
      </View>
    </View>
  </View>
</Modal>
`;
